base/twiceprecision.jl

```diff
@@ -196,8 +196,21 @@ ranges, you can set `nb = ceil(Int, log2(len-1))`.
 struct TwicePrecision{T}
     hi::T    # most significant bits
     lo::T    # least significant bits
+
+    function TwicePrecision{T}(hi::T, lo::T) where {T}
+        isconcretetype(T) || error("$T isn't a concrete type")
```

Comment:
This codepath can't be hit in practice - there are no concrete objects that have an abstract type as their type, so the `hi::T`/`lo::T` dispatch requirement can't be met.

Reply:
Not true:

```julia-repl
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.10.0-DEV.unknown (2023-05-06)
 _/ |\__'_|_|_|\__'_|  |  tp_disable_constructors/fc6dfb7999 (fork: 1 commits, 0 days)
|__/                   |

julia> Base.TwicePrecision{Any}(3, 0.3)
ERROR: Any isn't a concrete type
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] Base.TwicePrecision{Any}(hi::Int64, lo::Float64)
   @ Base ./twiceprecision.jl:201
 [3] top-level scope
   @ REPL[1]:1
```

Reply:
Thanks, that's true. I still think requiring this to be concrete is a bit odd.

Reply:
The reason that non-concrete type-parameters are usually allowed for container types in Julia is so the container's elements could be of different types, right?

In this case we don't want to allow elements of differing types, so there's no reason to allow non-concrete types. Even if I'm wrong the check can always be removed later, seeing as this is just an internal type.

That's my reasoning, anyway.

Reply:
Yes, that's true. The explicit check for `isconcretetype` just feels a bit iffy, is all. My guess is that the intention is to keep the `TwicePrecision{T}(a,b)` constructor around, right? I'm not sure whether that's actually needed (nothing *should* rely on it :sweat: ), but maybe something like this is better, instead of explicitly checking for concreteness of the type parameter:

```
julia> struct MyFoo{T}
           a::T
           b::T
           MyFoo(a::T, b::T) where T = new{T}(a,b)
       end

julia> MyFoo{T}(a::T, b::T) where T = MyFoo(a,b)

julia> MyFoo{Real}(1 ,2)
MyFoo{Int64}(1, 2)

julia> MyFoo{Real}(1, 2.0)
ERROR: MethodError: no method matching MyFoo(::Int64, ::Float64)

Closest candidates are:
  MyFoo(::T, ::T) where T
   @ Main REPL[1]:4

Stacktrace:
 [1] MyFoo{Real}(a::Int64, b::Float64)
   @ Main ./REPL[2]:1
 [2] top-level scope
   @ REPL[4]:1
```

That way, the constructor/API doesn't change, but we eliminate the case where the type parameter is specified to be different than the actual types of the passed in objects. Whether the `MethodError` is better than an explicit error message is debatable (though I'm inclined to prefer it, since it doesn't require reasoning about concreteness of types).

Reply:
Wow, this is a much prettier way to design the constructors!

Reply:
```
julia> MyFoo{Real}(1 ,2)
MyFoo{Int64}(1, 2)
```
That is horrifying and a big no-no.

Reply:
Then I guess the `isconcretetype` query has to stay, if we want to restrict the type to that extent.

Reply:
The more Julia-style way to write that query is probably that typeof(a)===typeof(b)

Reply:
That won't prevent `TwicePrecision{Real}(1,2)` though - for that you'd need `T === typeof(a) === typeof(b)`, which is the same as `isconcretetype` if I'm not mistaken.

Reply:
It is fairly abnormal to check the type parameter in that too